src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DynamicFiles/BackgroundDocumentGenerator.cs

```diff
@@ -143,114 +84,49 @@ public void Enqueue(IProjectSnapshot project, IDocumentSnapshot document)
             return;
         }
 
-        _dispatcher.AssertRunningOnDispatcher();
-
-        lock (Work)
+        if (Suppressed(project, document))
         {
-            if (Suppressed(project, document))
-            {
-                return;
-            }
-
-            // We only want to store the last 'seen' version of any given document. That way when we pick one to process
-            // it's always the best version to use.
-            Work[new DocumentKey(project.Key, document.FilePath.AssumeNotNull())] = (project, document);
-
-            StartWorker();
+            return;
         }
-    }
 
-    protected virtual void StartWorker()
-    {
-        // Access to the timer is protected by the lock in Enqueue and in Timer_Tick
-        // Timer will fire after a fixed delay, but only once.
-        _timer ??= NonCapturingTimer.Create(state => ((BackgroundDocumentGenerator)state).Timer_Tick(), this, Delay, Timeout.InfiniteTimeSpan);
+        _workQueue.AddWork((project, document));
     }
 
-    private void Timer_Tick()
+    protected virtual async ValueTask ProcessBatchAsync(ImmutableArray<(IProjectSnapshot, IDocumentSnapshot)> items, CancellationToken token)
     {
-        TimerTickAsync().Forget();
-    }
-
-    private async Task TimerTickAsync()
-    {
-        Assumes.NotNull(_timer);
-
-        try
+        foreach (var (project, document) in items.GetMostRecentUniqueItems(Comparer.Instance))
         {
-            // Timer is stopped.
-            _timer.Change(Timeout.Infinite, Timeout.Infinite);
-
-            OnStartingBackgroundWork();
-
-            KeyValuePair<DocumentKey, (IProjectSnapshot project, IDocumentSnapshot document)>[] work;
-            lock (Work)
+            if (_disposeTokenSource.IsCancellationRequested)
```

Comment:
Can/should this check `token` instead of `_disposeTokenSource`? I assume the `token` passed in is linked to that in some way anyway, since that is what was passed in to the queue constructor.

Reply:
These are equivalent. Every `CancellationToken` has a `CancellationTokenSource`, and `CancellationToken.IsCancellationRequested` just delegates to a token's `CancellationTokenSource`.

https://github.com/dotnet/runtime/blob/4039b4561595166f75577080ee784bc7f53e10a4/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationToken.cs#L65

Reply:
Isn't their equivalence dependent on an implementation detail of the AsyncBatchingWorkQueue though, that it doesn't have its own cancellation needs? I'm thinking of the CLaSP request queue, where the queue gets given a token to control it, but each piece of work is given a linked token that is made of that whole-queue token, and another token for just that piece of work. If the AsyncBatchingWorkQueue doesn't do that, and we don't think it ever will, then this can stay as is, it just feels odd to be in a method that is passed a cancellation token, and not used.

Reply:
Oops! I'm sorry, I was looking at a different part of the code. You are right of course! :smile: